pkg/operator/controller/externaldns/deployment.go

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package externaldnscontroller
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/google/go-cmp/cmp"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	operatorv1alpha1 "github.com/openshift/external-dns-operator/api/v1alpha1"
+	controller "github.com/openshift/external-dns-operator/pkg/operator/controller"
+)
+
+const (
+	metricsStartPort    = 7979
+	appNameLabel        = "app.kubernetes.io/name"
+	appInstanceLabel    = "app.kubernetes.io/instance"
+	masterNodeRoleLabel = "node-role.kubernetes.io/master"
+	osLabel             = "kubernetes.io/os"
+	linuxOS             = "linux"
+)
+
+// providerStringTable maps ExternalDNSProviderType values from the
+// ExternalDNS operator API to the provider string argument expected by ExternalDNS.
+var providerStringTable = map[operatorv1alpha1.ExternalDNSProviderType]string{
+	operatorv1alpha1.ProviderTypeAWS:      "aws",
+	operatorv1alpha1.ProviderTypeGCP:      "google",
+	operatorv1alpha1.ProviderTypeAzure:    "azure",
+	operatorv1alpha1.ProviderTypeBlueCat:  "bluecat",
+	operatorv1alpha1.ProviderTypeInfoblox: "infoblox",
+}
+
+// sourceStringTable maps ExternalDNSSourceType values from the
+// ExternalDNS operator API to the source string argument expected by ExternalDNS.
+var sourceStringTable = map[operatorv1alpha1.ExternalDNSSourceType]string{
+	operatorv1alpha1.SourceTypeRoute:   "openshift-route",
+	operatorv1alpha1.SourceTypeService: "service",
+	// TODO: Add CRD source support
+}
+
+// ensureExternalDNSDeployment ensures that the externalDNS deployment exists.
+// Returns a boolean if the deployment exists, and an error when relevant.
+func (r *reconciler) ensureExternalDNSDeployment(ctx context.Context, namespace, image string, serviceAccount *corev1.ServiceAccount, externalDNS *operatorv1alpha1.ExternalDNS) (bool, *appsv1.Deployment, error) {
+	desired, err := desiredExternalDNSDeployment(namespace, image, serviceAccount, externalDNS)
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to build externalDNS deployment: %w", err)
+	}
+
+	exist, current, err := r.currentExternalDNSDeployment(ctx, namespace, externalDNS)
```

Comment:
Instead of passing `namespace` and `externalDNS` to `currentExternalDNSDeployment` and other functions which then need to construct the namespaced name, it would be simpler to compute a `types.NamespacedName` value at the top of `ensureExternalDNSDeployment` and pass it down and use it throughout this function and its callees.  (BTW, a `types.NamespacedName` value is stringified as `namespace/name`, so it is convenient to use in error messages and log messages as well.)

Reply:
Done, for service account and namespace files too